### ISSUE TITLE
`Static` returning `never` for `Type.Intersect`ed fields

### ISSUE THREAD

**Issue by jonahsnider** [sinclairzx81/typebox#102]
Fields are `never` but should be `number | string`. The bug appears to be in either `Static` or `Type.Intersect`.

```ts
import { Static, Type } from '@sinclair/typebox';

const NumberObject = Type.Object({
	a: Type.Number(),
});
type NumberObject = Static<typeof NumberObject>;

const StringObject = Type.Object({
	a: Type.String(),
});
type StringObject = Static<typeof StringObject>;

const BaseObject = Type.Object({
	b: Type.String(),
});

const NumberOrStringObject = Type.Union([NumberObject, StringObject]);
type NumberOrStringObject = Static<typeof NumberOrStringObject>;

const test1: NumberOrStringObject = { a: 1 };
const test2: NumberOrStringObject = { a: '1' };

export const Combined = Type.Intersect([BaseObject, NumberOrStringObject]);

export type Combined = Static<typeof Combined>;

const test3: Combined = {
	a: 1,
	b: '1',
};
const test4: Combined = {
	a: '1',
	b: '1',
};

```

The generated schema seems valid:

```json
{
  "type": "object",
  "allOf": [
    { "type": "object", "properties": { "b": { "type": "string" } }, "required": ["b"] },
    {
      "anyOf": [
        { "type": "object", "properties": { "a": { "type": "number" } }, "required": ["a"] },
        { "type": "object", "properties": { "a": { "type": "string" } }, "required": ["a"] }
      ]
    }
  ]
}

```

<details><summary><b>Output</b></summary>

```ts
import { Type } from '@sinclair/typebox';
const NumberObject = Type.Object({
    a: Type.Number(),
}, {
    $id: 'RawStats',
    title: 'RawStats',
});
const StringObject = Type.Object({
    a: Type.String(),
});
const BaseObject = Type.Object({
    b: Type.String(),
});
const NumberOrStringObject = Type.Union([NumberObject, StringObject]);
const test1 = { a: 1 };
const test2 = { a: '1' };
export const Combined = Type.Intersect([BaseObject, NumberOrStringObject]);
const test3 = {
    a: 1,
    b: '1',
};
const test4 = {
    a: '1',
    b: '1',
};

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgZRgQxsAxgGjgFQE8wBTOAXzgDMoIQ4ByAAQGdgA7LAGzWCgHoYxEgCMIADwYBuALAAoeVgjsW8AHIBXECJJQA8iIBWJLPAC8BYQDoDx0wAp5ASATOnaAFyXSVzdt32AJQ4zuQhci5uACTAACZeDABKaADuqBgsDOFOTpgwXCQJyWnoMJnZYfKBsnJCpHB+OvpGJuYopdgAPHUkEFQNWk22rQB8NYrKqu1QHADmw6ZwFkQ+CzD2rhGe3iRWqDPss0Hh5NXyPdNza0vtGF09fZeHa2PyEyrwAEJoLCTXy9Y1htnCIvCtdvs5sd5KdxnIlB8Bv59FBIc8WosAT4AKrsYDKewAbUaujWeDR8wxMAAumdasIkUNUTADpS7G10pgsN1hI8SSiKS84QipjASKoAIxefl6Zms-6IODbCUUGoi+Bi1QAJmlg1JcquVJuSG2DAlDFVbzkJHEkFgcHVcAAwnQRBwSLEbuCrABJdhiqC-ByE76-MmM-WCqm0uE2u0ahku7Tuz0WTn3Xn9JNu9ge14KeGTDXimAAZi82ZTxucyuyoMY5pOaqLcE1MAALBXXVWLJt3AlGyCB1kYTUgA)

**Comment by sinclairzx81** [comment]
@jonahsnider Hi, thanks for posting this issue. As a workaround, you can define `NumberOrStringObject` as the following. See [Playground Link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgZRgQxsAxgGjgFQE8wBTOAXzgDMoIQ4ByAAQGdgA7LAGzWCgHoYxEgCMIADwYBuALAAoeVgjsW8AHIBXECJJQA8iIBWJLPAC8BYQDoDx0wAoE8gJBoAXJdJXN23fYCUOPLk-rJyQqRwPjr6RibmKOiYWAA8ESQQVFFaMbbxAHxhisqqiVAcAOZ5pnAWRF7VMI4u7p4kVqjl7BUBQXIhYelllY21iRjYacKZw92NhfLFKvAAQmgsJKN11o3Ncs4iHvXtnZW9waGLcvz8cErL2b76UKdzcTXbXgCq7MDK9gBtaK6Rp4V5Vd4wAC6lzk91KwOe4K2bRskL2cExcFaxysPz+7EBuPBvVRiICMIu8iGiL0LxgXQhdgSqAmqXSM1p9MZ8yKcJK8BgJFUAEYPFzkZCxkhWiKKGF4YLhTAAEzinIg7kjKUWGUeBgihjyq4kcSQWB3AVwADCdBEHBIABMxriAJLsIVQDYOAFrDagx65LVvZkwvmm82C4Q2u0O50WVnJKakGa27RxhYKfkPIWqADMHjT9vYTulLQ8Ir6B31hr65AVVtzMAALIXYyX44hy4xay5Dj2GHWpEA)

```typescript
// const NumberOrStringObject = Type.Union([NumberObject, StringObject]);
const NumberOrStringObject = Type.Object({
    a: Type.Union([Type.String(), Type.Number()])
})
type NumberOrStringObject = Static<typeof NumberOrStringObject>;
```
This problem relates to how TypeBox needs to statically infer intersections from variadic tuples. TypeBox first maps each type into a `union` and then converts the union to an intersection. At the heart of all this is the type `UnionToIntersect<T>` type. The following code demonstrates exactly where this problem arises. You can test [here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVQHbAqgKhAkqmAUygGdCBjeAXzgDMoIQ4ByAAROFXIBsBDYKAHoYATzCEARhAAezANwBYAFDLBguv24AuOLx3MSMKJwDmzXagAmLVAFcQE4uYgA3YnzBwSEOJ1rFdEhtCNyhmZVFxOABhOABeFHRMHHwiUgoYAB4kPS8jUzgaAB9EXR07BwDi0okdQ2NUE0KAPjhlVXUIAGt9etNzADIbe0cwuEtMZnhXd15PbzgoQm9uN0C4QmlxSkJLDrhIwhj4xIxsPAJiMkpssryGppKK0cK4EqRa+4KqZvalWlsXBgZwOyxgAAoXLxuLZCDpogBKRDKOCouBQmGEAB0vDgajgqBCxGUVCAA)

```typescript
import { UnionToIntersect } from '@sinclair/typebox';

// fail: a: 'string' and 'number' overlap so infer as 'never'
type C = UnionToIntersect<{ a: string } | { a: number } | { b: string }> 

// ok: a: 'string' | 'number' are union first don't overlap once intersected
// type C = UnionToIntersect<{ a: string | number } | { b: string }>

function test(value: C) {
    value.a // test here
}
```
Unfortunately, to resolve this issue, one would need to resolve the `UnionToIntersect<T>` type; and as of current (and to the best of my knowledge), there isn't a better way to produce an intersection from variadic tuple (note that `UnionToIntersect<T>` is very common, you can read a bit about it [here](https://fettblog.eu/typescript-union-to-intersection/).)

I will keep this issue open for a day or two and have a read up on latest enhancements to the compiler. I'm pretty sure this is non resolvable though, so would recommend the above workaround to express these kind of unions.

Hope this helps
S

**Comment by sinclairzx81** [comment]
@jonahsnider Hey, just an quick update. I might actually flag this issue as a future enhancement. There's been a few things I've found while looking around that I'm quite interested in digging a bit deeper into (specifically related to recursive type inference which opens up quite a few options). Such enhancements would implicate static inference across most of TypeBox, so there would be a potential to address this issue in the process. 

I can't give an estimate on when I'd be able to resolve this particular inference issue however (which would involve evaluating the Union Intersection as some form of expression with evaluation precedence (i.e `(A | B) & C` instead of `A | (B & C)` where `&` takes precedence over `|`), but I am quite mindful of it. For the time being my recommendation would be to keep on with the original workaround until I've had a bit of time to sit down and look at things. 

Thanks again for submitting the issue. These cases do help improve the codebase. 
S

**Comment by SimenB** [comment]
I have the same issue (I think). However, I'm trying to migrate from `runtypes` to `@sinclair/typebox` which indicates what I want can be done by TS, so maybe you can pick up some inspiration from there?

[Playground](https://www.typescriptlang.org/play?ssl=43&ssc=2&pln=23&pc=2#code/JYWwDg9gTgLgBAKjgQwM51nAZlCI4DkUArgHYwCeYApqgQNwCwAUKJJgN5wAqV1ANHADKMZDGABjOAF9sufAQACqYKQkAbZMCgB6SjQBGEAB4MWLCRFKp4sbgAtVAczgBeDDBZw4AOgBK1JZQACYAFFxYEBAGyFAAXB4+IlDOMgCUXr4AgqRhmd6wPgCqpMBWofneiQFBYRHA1OrBCYUAMsAw1FDI6qEEPVbUBGnp-JUFMP6B0HXYDU0tk+2d3b0E9mgAYinUucOCOA25i0kwKaQu0mljzFVw15lpTMws+tRwfjC8NG6JImKSAA8bwgWA8DmcAD5nixqMZ2PBLNZbDAAIwJT7fd7uDgsACQkWisQSBFQeHeNnOTgINwJ82ahAGpCGN2ksPh0ERVhsHgATBivnxfrjmASojF4oQySB3hAYPYunBKc4afisPSSRtUNsjsFVaLDrsGQQAJrICgEFhsl7MJE8t5GYwQi6-LE+ACS5C6qECMFCAG1Mm6APIGABWvvC2HFxJ4fFOVNCIyuN28bpKZVIAcqIfDka4uHU6moCTdyy6PT6TKGyYetyquYjEj9BYgRZLcZoPnLqz6Wp1Rv2cl1pfjyWcSdGmQAunXZzDmG9O9RHVjfv9xBJgXxQXAHSZnU5oeZmHCEXA7fB98Z0cvV0KcWqY5LSeSlWcVbT1Y1jdX9daz05C9uSvPhHX5O8DwfOARTFIkX2lWV5UVZULn1Okf01LYdj2L8cONM0LStIA)

![image](https://user-images.githubusercontent.com/1404810/136520530-416e9b82-5e94-468f-a265-ab3af9885394.png)

It seems to me like it does the precedence thing you mentioned, so I guess you're already down this track 😀 But do you have a suggested workaround I could use? I don't think the separate union you suggested will work in my case since depending on its literal value there's another field there

**Comment by sinclairzx81** [comment]
@jonahsnider @SimenB Hi, Have just published an update on `@sinclair/typebox@0.21.2` that resolves this issue. I've made mention in the changelog about the update which you can read about below. 

This work was undertaken to align with various changes in TS 4.5.

ChangeLog [link](https://github.com/sinclairzx81/typebox/blob/master/changelog.md)

TypeScript Playground [link](https://www.typescriptlang.org/play?ts=4.5.0-beta#code/JYWwDg9gTgLgBAbzgZRgQxsAxgGjgFQE8wBTOAXzgDMoIQ4AiAAQGdgA7LAGzWCgHoYxEgCMIADwYBuALAAoeVgjsW8AIJwAvAWEA6APIiAViSwwAFEjQAuHaV2ooHAObmAlBTeLlquACEtOxIDY1MLJBFbIntHF3dPbxV4AGFA6ODDEzNLOCwovVj2Vw9yLwU5JSSCNL0ASXYYEigWMPMAbTU8dN0AVXZgZXa-PGSAXTdx+SFSau1UDGwAHmmSCCoCAD55eX5+OBXZxDgbOFUnIoo4ADI4HMjTmHPnS4AfI7yHp4TyoA)

Thanks for reporting this issue and apologies for the delay in getting this update ready. These changes were only really feasible following the optimization updates to align with TS 4.5.

Will close off this issue in a couple of days, but let me know if you run into any problems.
Cheers
S

**Comment by SimenB** [comment]
Thanks! My real code is working correctly now 👍
